Source/WebCore/page/FrameView.cpp

```diff
@@ -0,0 +1,345 @@
+
+#include "Chrome.h"
+#include "ChromeClient.h"
+#include "FocusController.h"
+#include "Frame.h"
+#include "HTMLFrameOwnerElement.h"
+#include "Page.h"
+#include "RenderElement.h"
+#include "RenderLayer.h"
+#include "RenderLayerScrollableArea.h"
+#include "RenderWidget.h"
+
+namespace WebCore {
+
+int FrameView::headerHeight() const
+{
+    Ref frame = this->frame();
+    if (!frame->isMainFrame())
+        return 0;
+    Page* page = frame->page();
+    return page ? page->headerHeight() : 0;
+}
+
+int FrameView::footerHeight() const
+{
+    Ref frame = this->frame();
+    if (!frame->isMainFrame())
+        return 0;
+    Page* page = frame->page();
+    return page ? page->footerHeight() : 0;
+}
+
+float FrameView::topContentInset(TopContentInsetType contentInsetTypeToReturn) const
+{
+    if (platformWidget() && contentInsetTypeToReturn == TopContentInsetType::WebCoreOrPlatformContentInset)
+        return platformTopContentInset();
+
+    Ref frame = this->frame();
+    if (!frame->isMainFrame())
+        return 0;
+
+    Page* page = frame->page();
+    return page ? page->topContentInset() : 0;
+}
+
+float FrameView::visibleContentScaleFactor() const
+{
+    Ref frame = this->frame();
+    if (!frame->isMainFrame())
+        return 1;
+
+    Page* page = frame->page();
+    // FIXME: This !delegatesScaling() is confusing, and the opposite behavior to Frame::frameScaleFactor().
+    // This function should probably be renamed to delegatedPageScaleFactor().
+    if (!page || !page->delegatesScaling())
+        return 1;
+
+    return page->pageScaleFactor();
+}
+
+bool FrameView::isActive() const
+{
+    Page* page = frame().page();
+    return page && page->focusController().isActive();
+}
+
+ScrollableArea* FrameView::enclosingScrollableArea() const
+{
+    Ref frame = this->frame();
+    if (frame->isMainFrame())
+        return nullptr;
+
+    auto* ownerElement = frame->ownerElement();
+    if (!ownerElement)
+        return nullptr;
+
+    auto* ownerRenderer = ownerElement->renderer();
+    if (!ownerRenderer)
+        return nullptr;
+
+    auto* layer = ownerRenderer->enclosingLayer();
+    if (!layer)
+        return nullptr;
+
+    auto* enclosingScrollableLayer = layer->enclosingScrollableLayer(IncludeSelfOrNot::IncludeSelf, CrossFrameBoundaries::No);
+    if (!enclosingScrollableLayer)
+        return nullptr;
+
+    return enclosingScrollableLayer->scrollableArea();
+}
+
+void FrameView::invalidateRect(const IntRect& rect)
+{
+    Ref frame = this->frame();
+    if (!parent()) {
+        if (auto* page = frame->page())
+            page->chrome().invalidateContentsAndRootView(rect);
+        return;
+    }
+
+    auto* renderer = frame->ownerRenderer();
+    if (!renderer)
+        return;
+
+    IntRect repaintRect = rect;
+    repaintRect.moveBy(roundedIntPoint(renderer->contentBoxLocation()));
+    renderer->repaintRectangle(repaintRect);
+}
+
+bool FrameView::forceUpdateScrollbarsOnMainThreadForPerformanceTesting() const
```

Comment:
Probably in a different PR, but this should probably be renamed to scrollingPerformanceTestingEnabled.  It looks like it used to do something, but now it just returns that boolean from the page.